File: main/src/telemetry/otlpReceiver.ts
import express, { Express, Request, Response } from 'express';
import { Server } from 'http';
import { Logger } from '../utils/logger';
import { EventEmitter } from 'events';
import * as zlib from 'zlib';

interface OTLPMetric {
  name: string;
  gauge?: {
    dataPoints: Array<{
      asInt?: number;
      asDouble?: number;
      attributes?: Array<{
        key: string;
        value: {
          stringValue?: string;
          intValue?: number;
        };
      }>;
    }>;
  };
  sum?: {
    dataPoints: Array<{
      asInt?: number;
      asDouble?: number;
      attributes?: Array<{
        key: string;
        value: {
          stringValue?: string;
          intValue?: number;
        };
      }>;
    }>;
  };
  histogram?: {
    dataPoints: Array<{
      count?: number;
      sum?: number;
      attributes?: Array<{
        key: string;
        value: {
          stringValue?: string;
          intValue?: number;
        };
      }>;
    }>;
  };
}

interface OTLPResourceMetrics {
  resource?: {
    attributes?: Array<{
      key: string;
      value: {
        stringValue?: string;
        intValue?: number;
      };
    }>;
  };
  scopeMetrics?: Array<{
    scope?: {
      name: string;
      version?: string;
    };
    metrics?: OTLPMetric[];
  }>;
}

interface OTLPMetricsPayload {
  resourceMetrics?: OTLPResourceMetrics[];
}

interface OTLPLogRecord {
  timeUnixNano?: string;
  severityNumber?: number;
  severityText?: string;
  body?: {
    stringValue?: string;
    kvlistValue?: {
      values?: Array<{
        key: string;
        value: {
          stringValue?: string;
          intValue?: string;
          doubleValue?: number;
          boolValue?: boolean;
        };
      }>;
    };
  };
  attributes?: Array<{
    key: string;
    value: {
      stringValue?: string;
      intValue?: string;
      doubleValue?: number;
      boolValue?: boolean;
    };
  }>;
}

interface OTLPLogsPayload {
  resourceLogs?: Array<{
    resource?: {
      attributes?: Array<{
        key: string;
        value: {
          stringValue?: string;
        };
      }>;
    };
    scopeLogs?: Array<{
      scope?: {
        name: string;
        version?: string;
      };
      logRecords?: OTLPLogRecord[];
    }>;
  }>;
}

export interface TelemetryStatus {
  sessions: number;
  totalTokens: number;
  inputTokens: number;
  outputTokens: number;
  cacheReadTokens?: number;
  cacheCreationTokens?: number;
  totalCost: number;
  linesAdded: number;
  linesRemoved: number;
  activeTimeMs: number;
  toolUsage: Record<string, number>;
  lastUpdate: string | null;
  currentSessionId: string | null;
}

export class OTLPReceiver extends EventEmitter {
  private app: Express;
  private server?: Server;
  private port: number;
  private status: TelemetryStatus;
  private logger: Logger;

  constructor(port: number = 4318, logger: Logger) {
    super();
    this.port = port;
    this.logger = logger;
    this.app = express();
    this.status = {
      sessions: 0,
      totalTokens: 0,
      inputTokens: 0,
      outputTokens: 0,
      cacheReadTokens: 0,
      cacheCreationTokens: 0,
      totalCost: 0,
      linesAdded: 0,
      linesRemoved: 0,
      activeTimeMs: 0,
      toolUsage: {},
      lastUpdate: null,
      currentSessionId: null,
    };

    this.setupRoutes();
  }

  private setupRoutes(): void {
    // Middleware to handle gzip decompression
    this.app.use((req, res, next) => {
      if (req.headers['content-encoding'] === 'gzip') {
        const chunks: Buffer[] = [];
        req.on('data', (chunk) => chunks.push(chunk));
        req.on('end', () => {
          const buffer = Buffer.concat(chunks);
          zlib.gunzip(buffer, (err, decompressed) => {
            if (err) {
              this.logger.error('Failed to decompress gzip data:', err);
              res.status(400).json({ error: 'Invalid gzip data' });
              return;
            }
            try {
              req.body = JSON.parse(decompressed.toString());
              next();
            } catch (parseErr) {
              this.logger.error('Failed to parse decompressed data:', parseErr as Error);
              res.status(400).json({ error: 'Invalid JSON data' });
            }
          });
        });
      } else {
        next();
      }
    });
    
    this.app.use(express.json({ limit: '10mb' }));

    // OTLP HTTP endpoint for metrics
    this.app.post('/v1/metrics', (req: Request, res: Response) => {
      try {
        // this.logger.info(`Received OTLP metrics request from ${req.ip}`);
        // this.logger.info(`Content-Type: ${req.headers['content-type']}`);
        // this.logger.info(`Body: ${JSON.stringify(req.body).substring(0, 200)}...`);
        // this.processMetrics(req.body);
        res.status(200).json({ status: 'ok' });
      } catch (error) {
        this.logger.error('Error processing OTLP metrics:', error as Error);
        res.status(500).json({ error: error instanceof Error ? error.message : 'Unknown error' });
      }
    });

    // OTLP HTTP endpoint for traces
    this.app.post('/v1/traces', (req: Request, res: Response) => {
      try {
        this.logger.info('Received OTLP traces');
        // For now, we're focused on metrics, but traces could be processed here
        res.status(200).json({ status: 'ok' });
      } catch (error) {
        this.logger.error('Error processing OTLP traces:', error as Error);
        res.status(500).json({ error: error instanceof Error ? error.message : 'Unknown error' });
      }
    });

    // OTLP HTTP endpoint for logs
    this.app.post('/v1/logs', (req: Request, res: Response) => {
      try {
        // this.logger.info(`Received OTLP logs request from ${req.ip}`);
        // this.logger.info(`Content-Type: ${req.headers['content-type']}`);
        // this.logger.info(`Content-Encoding: ${req.headers['content-encoding']}`);
        
        // Process log data if needed
        this.processLogs(req.body);
        
        res.status(200).json({ status: 'ok' });
      } catch (error) {
        this.logger.error('Error processing OTLP logs:', error as Error);
        res.status(500).json({ error: error instanceof Error ? error.message : 'Unknown error' });
      }
    });

    // Status endpoint for querying current metrics
    this.app.get('/status', (req: Request, res: Response) => {
      res.json(this.status);
    });

    // Health check
    this.app.get('/health', (req: Request, res: Response) => {
      res.json({ status: 'healthy', uptime: process.uptime() });
    });

    // Catch-all for debugging - must be last
    this.app.use((req: Request, res: Response) => {
      this.logger.info(`Unhandled request: ${req.method} ${req.path}`);
      this.logger.info(`Headers: ${JSON.stringify(req.headers)}`);
      res.status(404).json({ error: 'Not found' });
    });
  }

  private processLogs(payload: OTLPLogsPayload): void {
    if (!payload.resourceLogs) {
      this.logger.warn('No resourceLogs in OTLP logs payload');
      return;
    }

    let logCount = 0;
    let telemetryUpdated = false;

    payload.resourceLogs.forEach(rl => {
      // Extract resource attributes if needed
      const resourceAttrs = this.extractAttributes(rl.resource?.attributes || []);
      
      rl.scopeLogs?.forEach(sl => {
        const scopeName = sl.scope?.name || 'unknown';
        // this.logger.info(`Processing logs from scope: ${scopeName}`);
        
        sl.logRecords?.forEach(logRecord => {
          logCount++;
          
          // Extract log severity (currently unused)
          // const severity = logRecord.severityText || `SEVERITY_${logRecord.severityNumber}` || 'UNKNOWN';
          // logSummary[severity] = (logSummary[severity] || 0) + 1;
          
          // Extract log body
          let logMessage = '';
          if (logRecord.body?.stringValue) {
            logMessage = logRecord.body.stringValue;
          } else if (logRecord.body?.kvlistValue?.values) {
            // Handle structured logs
            const kvPairs = logRecord.body.kvlistValue.values.map(kv => 
              `${kv.key}=${kv.value.stringValue || kv.value.intValue || kv.value.doubleValue || kv.value.boolValue}`
            );
            logMessage = kvPairs.join(', ');
          }
          
          // Extract attributes
          const attrs = this.extractAttributes(logRecord.attributes || []);
          
          // Process API request logs to extract accurate token and cost data
          if (attrs['event.name'] === 'api_request' || attrs['event.name'] === 'user_prompt') {
            const inputTokens = parseInt(attrs['input_tokens'] || '0');
            const outputTokens = parseInt(attrs['output_tokens'] || '0');
            const cacheReadTokens = parseInt(attrs['cache_read_tokens'] || '0');
            const cacheCreationTokens = parseInt(attrs['cache_creation_tokens'] || '0');
            const cost = parseFloat(attrs['cost_usd'] || '0');
            
            // Update our telemetry status with the actual values from logs
            this.status.inputTokens += inputTokens;
            this.status.outputTokens += outputTokens;
            this.status.totalTokens = this.status.inputTokens + this.status.outputTokens;
            this.status.totalCost += cost;
            
            // Track cache tokens separately if needed
            if (!this.status.cacheReadTokens) this.status.cacheReadTokens = 0;
            if (!this.status.cacheCreationTokens) this.status.cacheCreationTokens = 0;
            this.status.cacheReadTokens += cacheReadTokens;
            this.status.cacheCreationTokens += cacheCreationTokens;
            
            telemetryUpdated = true;
            
            this.logger.info(`[API Request] Tokens: ${inputTokens} in, ${outputTokens} out, Cache: ${cacheReadTokens} read, ${cacheCreationTokens} created, Cost: $${cost.toFixed(2)}`);
          }
          
          // Log at appropriate level
          const timestamp = logRecord.timeUnixNano ? 
            new Date(parseInt(logRecord.timeUnixNano) / 1_000_000).toISOString() : 
            new Date().toISOString();
          
          // this.logger.info(`[OTLP Log] ${timestamp} [${severity}] ${logMessage} ${JSON.stringify(attrs)}`);
        });
      });
    });

    // this.logger.info(`Processed ${logCount} log records. Summary: ${JSON.stringify(logSummary)}`);
    
    // If we updated telemetry from logs, emit the update
    if (telemetryUpdated) {
      this.status.lastUpdate = new Date().toISOString();
      
      // Emit telemetry update with data from logs
      this.emit('telemetry-update', {
        metrics: {
          sessionCount: this.status.sessions,
          linesOfCodeModified: this.status.linesAdded + this.status.linesRemoved,
          apiCost: this.status.totalCost,
          tokenUsage: {
            input: this.status.inputTokens,
            output: this.status.outputTokens,
            total: this.status.totalTokens,
            cacheRead: this.status.cacheReadTokens,
            cacheCreation: this.status.cacheCreationTokens,
          },
          activeTimeMs: this.status.activeTimeMs,
          toolDecisions: this.status.toolUsage,
        },
        traces: {
          spans: [],
        },
        lastUpdate: this.status.lastUpdate,
      });
      
      this.logger.info(`Updated telemetry from logs - Total tokens: ${this.status.totalTokens}, Total cost: $${this.status.totalCost.toFixed(2)}`);
    }
    
    // Emit event for log processing
    this.emit('logs-received', {
      count: logCount,
      timestamp: new Date().toISOString()
    });
  }

  private processMetrics(payload: OTLPMetricsPayload): void {
    if (!payload.resourceMetrics) {
      this.logger.warn('No resourceMetrics in OTLP payload');
      return;
    }

    payload.resourceMetrics.forEach(rm => {
      // Extract resource attributes if needed
      const resourceAttrs = this.extractAttributes(rm.resource?.attributes || []);
      
      rm.scopeMetrics?.forEach(sm => {
        const scopeName = sm.scope?.name || 'unknown';
        this.logger.info(`Processing metrics from scope: ${scopeName}`);
        
        sm.metrics?.forEach(metric => {
          this.updateStatus(metric);
        });
      });
    });

    this.status.lastUpdate = new Date().toISOString();
    
    // Emit event with updated telemetry data

    this.emit('telemetry-update', {
      metrics: {
        sessionCount: this.status.sessions,
        linesOfCodeModified: this.status.linesAdded + this.status.linesRemoved,
        apiCost: this.status.totalCost,
        tokenUsage: {
          input: this.status.inputTokens,
          output: this.status.outputTokens,
          total: this.status.totalTokens,
        },
        activeTimeMs: this.status.activeTimeMs,
        toolDecisions: this.status.toolUsage,
      },
      traces: {
        spans: [], // Could be populated from trace data
      },
      lastUpdate: this.status.lastUpdate,
    });

  }

  private updateStatus(metric: OTLPMetric): void {
    const name = metric.name;
    this.logger.info(`Processing metric: ${name}`);

    switch (name) {
      case 'claude.sessions.total':
      case 'claude_code.session.count':
        this.status.sessions = this.getMetricValue(metric);
        this.status.currentSessionId = this.getAttribute(metric, 'session.id') || 
                                       this.getAttribute(metric, 'session_id') || null;
        break;

      case 'claude.tokens.total':
      case 'claude_code.token.usage':
        // NOTE: Token counts from metrics appear to be less accurate than logs
        // We now get more accurate token counts from the logs API request events
        const tokens = this.getMetricValue(metric);
        const tokenType = this.getAttribute(metric, 'type') || 
                         this.getAttribute(metric, 'token_type');
        
        // Only update if we haven't received log data yet
        if (this.status.totalTokens === 0) {
          if (tokenType === 'input') {
            this.status.inputTokens += tokens;
          } else if (tokenType === 'output') {
            this.status.outputTokens += tokens;
          }
          this.status.totalTokens = this.status.inputTokens + this.status.outputTokens;
        }
        break;

      case 'claude.api_calls.total':
      case 'claude_code.api_calls.count':
        // Could track API call counts here
        break;

      case 'claude.api_cost.total':
      case 'claude_code.cost.usage':
        // NOTE: Cost from metrics appears to be less accurate than logs
        // Only update if we haven't received log data yet
        if (this.status.totalCost === 0) {
          this.status.totalCost += this.getMetricValue(metric);
        }
        break;

      case 'claude.lines_modified.total':
      case 'claude_code.lines_of_code.count':
        const lines = this.getMetricValue(metric);
        const lineType = this.getAttribute(metric, 'type') || 
                        this.getAttribute(metric, 'operation');
        
        if (lineType === 'added' || lineType === 'add') {
          this.status.linesAdded += lines;
        } else if (lineType === 'removed' || lineType === 'remove') {
          this.status.linesRemoved += lines;
        }
        break;

      case 'claude.active_time.duration':
      case 'claude_code.active_time.ms':
        this.status.activeTimeMs = this.getMetricValue(metric);
        break;

      case 'claude.tool_usage.total':
      case 'claude_code.tool_usage.count':
        const toolName = this.getAttribute(metric, 'tool') || 
                        this.getAttribute(metric, 'tool_name') || 
                        'unknown';
        const count = this.getMetricValue(metric);
        
        if (!this.status.toolUsage[toolName]) {
          this.status.toolUsage[toolName] = 0;
        }
        this.status.toolUsage[toolName] += count;
        break;

      default:
        this.logger.info(`Unhandled metric: ${name}`);
    }
  }

  private getMetricValue(metric: OTLPMetric): number {
    // Handle different metric types
    let dataPoints: any[] = [];
    
    if (metric.sum?.dataPoints) {
      dataPoints = metric.sum.dataPoints;
    } else if (metric.gauge?.dataPoints) {
      dataPoints = metric.gauge.dataPoints;
    } else if (metric.histogram?.dataPoints) {
      // For histograms, we might want the sum or count
      return metric.histogram.dataPoints[0]?.sum || 
             metric.histogram.dataPoints[0]?.count || 0;
    }

    if (dataPoints.length > 0) {
      const point = dataPoints[0];
      return point.asInt || point.asDouble || 0;
    }

    return 0;
  }

  private getAttribute(metric: OTLPMetric, attrName: string): string | undefined {
    const dataPoints = metric.sum?.dataPoints || 
                      metric.gauge?.dataPoints || 
                      metric.histogram?.dataPoints || [];
    
    if (dataPoints.length > 0) {
      const attributes = dataPoints[0].attributes || [];
      const attr = attributes.find(a => a.key === attrName);
      return attr?.value?.stringValue || 
             (attr?.value?.intValue !== undefined ? String(attr.value.intValue) : undefined);
    }
    
    return undefined;
  }

  private extractAttributes(attributes: Array<{ key: string; value: any }>): Record<string, any> {
    const result: Record<string, any> = {};
    
    attributes.forEach(attr => {
      result[attr.key] = attr.value.stringValue || 
                        attr.value.intValue || 
                        attr.value.boolValue || 
                        attr.value;
    });
    
    return result;
  }

  start(): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        this.server = this.app.listen(this.port, () => {
          this.logger.info(`OTLP Receiver running on http://localhost:${this.port}`);
          this.logger.info(`Metrics endpoint: http://localhost:${this.port}/v1/metrics`);
          this.logger.info(`Status API: http://localhost:${this.port}/status`);
          resolve();
        });

        this.server.on('error', (error: any) => {
          if (error.code === 'EADDRINUSE') {
            this.logger.warn(`Port ${this.port} is in use, OTLP receiver not started`);
            resolve(); // Don't fail the app if port is in use
          } else {
            reject(error);
          }
        });
      } catch (error) {
        reject(error);
      }
    });
  }

  stop(): Promise<void> {
    return new Promise((resolve) => {
      if (this.server) {
        this.server.close(() => {
          this.logger.info('OTLP Receiver stopped');
          resolve();
        });
      } else {
        resolve();
      }
    });
  }

  getStatus(): TelemetryStatus {
    return { ...this.status };
  }

  resetStatus(): void {
    this.status = {
      sessions: 0,
      totalTokens: 0,
      inputTokens: 0,
      outputTokens: 0,
      cacheReadTokens: 0,
      cacheCreationTokens: 0,
      totalCost: 0,
      linesAdded: 0,
      linesRemoved: 0,
      activeTimeMs: 0,
      toolUsage: {},
      lastUpdate: null,
      currentSessionId: null,
    };
  }
}

File: src-tauri/src/telemetry/otlp_receiver.rs
use axum::{
    extract::State,
    http::StatusCode,
    response::Json,
    routing::{get, post},
    Router,
};
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use tokio::sync::{broadcast, RwLock};
use tower_http::cors::CorsLayer;
use tracing::{error, info};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TelemetryData {
    pub tokens_input: u64,
    pub tokens_output: u64,
    pub tokens_total: u64,
    pub cache_read_tokens: Option<u64>,
    pub cache_creation_tokens: Option<u64>,
    pub cost_usd: f64,
    pub tool_usage: std::collections::HashMap<String, u64>,
    pub active_time_ms: u64,
    pub last_update: Option<String>,
}

impl Default for TelemetryData {
    fn default() -> Self {
        Self {
            tokens_input: 0,
            tokens_output: 0,
            tokens_total: 0,
            cache_read_tokens: None,
            cache_creation_tokens: None,
            cost_usd: 0.0,
            tool_usage: std::collections::HashMap::new(),
            active_time_ms: 0,
            last_update: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GenerationProgress {
    pub stage: String,
    pub message: String,
    pub percentage: u8,
    pub telemetry: Option<TelemetryData>,
}

#[derive(Debug, Deserialize)]
struct OtlpMetric {
    name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    gauge: Option<MetricPoints>,
    #[serde(skip_serializing_if = "Option::is_none")]
    sum: Option<MetricPoints>,
}

#[derive(Debug, Deserialize)]
struct MetricPoints {
    #[serde(rename = "dataPoints")]
    data_points: Vec<DataPoint>,
}

#[derive(Debug, Deserialize)]
struct DataPoint {
    #[serde(rename = "asInt", skip_serializing_if = "Option::is_none")]
    as_int: Option<i64>,
    #[serde(rename = "asDouble", skip_serializing_if = "Option::is_none")]
    as_double: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    attributes: Option<Vec<Attribute>>,
}

#[derive(Debug, Deserialize)]
struct Attribute {
    key: String,
    value: AttributeValue,
}

#[derive(Debug, Deserialize)]
struct AttributeValue {
    #[serde(rename = "stringValue", skip_serializing_if = "Option::is_none")]
    string_value: Option<String>,
    #[serde(rename = "intValue", skip_serializing_if = "Option::is_none")]
    int_value: Option<i64>,
}

#[derive(Debug, Deserialize)]
struct OtlpPayload {
    #[serde(rename = "resourceMetrics")]
    resource_metrics: Vec<ResourceMetrics>,
}

#[derive(Debug, Deserialize)]
struct ResourceMetrics {
    #[serde(rename = "scopeMetrics")]
    scope_metrics: Vec<ScopeMetrics>,
}

#[derive(Debug, Deserialize)]
struct ScopeMetrics {
    metrics: Vec<OtlpMetric>,
}

#[derive(Debug, Deserialize)]
struct OtlpLogsPayload {
    #[serde(rename = "resourceLogs")]
    resource_logs: Vec<ResourceLogs>,
}

#[derive(Debug, Deserialize)]
struct ResourceLogs {
    #[serde(rename = "scopeLogs")]
    scope_logs: Vec<ScopeLogs>,
}

#[derive(Debug, Deserialize)]
struct ScopeLogs {
    #[serde(rename = "logRecords")]
    log_records: Vec<LogRecord>,
}

#[derive(Debug, Deserialize)]
struct LogRecord {
    #[allow(dead_code)]
    body: Option<LogBody>,
    attributes: Option<Vec<Attribute>>,
}

#[derive(Debug, Deserialize)]
struct LogBody {
    #[serde(rename = "stringValue")]
    #[allow(dead_code)]
    string_value: Option<String>,
}

#[derive(Clone)]
struct AppState {
    telemetry: Arc<RwLock<TelemetryData>>,
    progress_tx: broadcast::Sender<GenerationProgress>,
}

pub struct OtlpReceiver {
    port: u16,
    telemetry: Arc<RwLock<TelemetryData>>,
    progress_tx: broadcast::Sender<GenerationProgress>,
    server_handle: Option<tokio::task::JoinHandle<()>>,
}

impl OtlpReceiver {
    pub fn new(port: u16) -> Self {
        let (progress_tx, _) = broadcast::channel(100);
        Self {
            port,
            telemetry: Arc::new(RwLock::new(TelemetryData::default())),
            progress_tx,
            server_handle: None,
        }
    }

    pub fn port(&self) -> u16 {
        self.port
    }

    pub async fn start(&mut self) -> Result<(), String> {
        let state = AppState {
            telemetry: self.telemetry.clone(),
            progress_tx: self.progress_tx.clone(),
        };

        let app = Router::new()
            .route("/v1/metrics", post(handle_metrics))
            .route("/v1/logs", post(handle_logs))
            .route("/v1/traces", post(handle_traces))
            .route("/status", get(get_status))
            .route("/health", get(health_check))
            .layer(CorsLayer::permissive())
            .layer(tower_http::trace::TraceLayer::new_for_http()
                .on_request(|request: &axum::http::Request<_>, _span: &tracing::Span| {
                    //println!("OTLP: Incoming request: {} {}", request.method(), request.uri());
                    //println!("OTLP: Headers: {:?}", request.headers());
                }))
            .with_state(state);

        let addr = format!("127.0.0.1:{}", self.port);
        info!("Starting OTLP receiver on {}", addr);
        println!("OTLP: Starting receiver on {}", addr);
        //println!("OTLP: Test endpoints:");
        //println!("  curl http://localhost:{}/health", self.port);
        //println!("  curl http://localhost:{}/status", self.port);

        let listener = tokio::net::TcpListener::bind(&addr).await
            .map_err(|e| format!("Failed to bind to {}: {}", addr, e))?;
        
        let server = axum::serve(listener, app);
        
        let handle = tokio::spawn(async move {
            if let Err(e) = server.await {
                error!("OTLP server error: {}", e);
            }
        });

        self.server_handle = Some(handle);
        Ok(())
    }

    pub async fn stop(&mut self) {
        if let Some(handle) = self.server_handle.take() {
            handle.abort();
            info!("OTLP receiver stopped");
        }
    }

    pub fn subscribe(&self) -> broadcast::Receiver<GenerationProgress> {
        self.progress_tx.subscribe()
    }

    pub async fn get_telemetry(&self) -> TelemetryData {
        self.telemetry.read().await.clone()
    }

    pub async fn reset_telemetry(&self) {
        let mut telemetry = self.telemetry.write().await;
        *telemetry = TelemetryData::default();
        println!("OTLP: Telemetry data reset");
    }
}

async fn handle_metrics(
    State(_state): State<AppState>,
    Json(payload): Json<OtlpPayload>,
) -> StatusCode {    
    StatusCode::OK
}

async fn handle_logs(
    State(state): State<AppState>,
    Json(payload): Json<OtlpLogsPayload>,
) -> StatusCode {
    println!("OTLP: ====== Received logs request ======");
    //println!("OTLP: Timestamp: {}", chrono::Utc::now().to_rfc3339());
    let mut telemetry = state.telemetry.write().await;
    
    let mut _log_count = 0;
    for rl in payload.resource_logs {
        for sl in rl.scope_logs {
            //println!("OTLP: Processing {} log records", sl.log_records.len());
            for log in sl.log_records {
                _log_count += 1;
                if let Some(attrs) = log.attributes {
                //    println!("OTLP: Processing log with {} attributes", attrs.len());
                    process_log_attributes(&mut telemetry, attrs);
                }
            }
        }
    }
    
    //println!("OTLP: Processed {} log records total", _log_count);
    telemetry.last_update = Some(chrono::Utc::now().to_rfc3339());
    
    // Calculate progress based on telemetry

    println!("OTLP: After logs - tokens: {}, cost: ${:.3}", 
        telemetry.tokens_total, telemetry.cost_usd);
    
    // Send progress update if we got meaningful data
    if telemetry.tokens_total > 0 {
        let result = state.progress_tx.send(GenerationProgress {
            stage: "processing".to_string(),
            message: format!("Processing... ({} tokens)", telemetry.tokens_total),
            percentage: 50,
            telemetry: Some(telemetry.clone()),
        });
        
        match result {
            Ok(count) => {
            //    println!("OTLP: Progress update sent to {} receivers from logs", count);
                println!("OTLP: Sent progress from logs - tokens: {}", telemetry.tokens_total);
            },
            Err(e) => println!("OTLP: Failed to send progress update from logs: {:?}", e),
        }
    }
    
    StatusCode::OK
}

async fn handle_traces(
    State(_state): State<AppState>,
    Json(_payload): Json<serde_json::Value>,
) -> StatusCode {
    // We don't process traces for now
    StatusCode::OK
}

async fn get_status(State(state): State<AppState>) -> Json<TelemetryData> {
    Json(state.telemetry.read().await.clone())
}

async fn health_check() -> Json<serde_json::Value> {
    Json(serde_json::json!({
        "status": "healthy",
        "timestamp": chrono::Utc::now().to_rfc3339()
    }))
}

fn update_telemetry(telemetry: &mut TelemetryData, metric: &OtlpMetric) {
    let value = get_metric_value(metric);
    println!("OTLP: Metric '{}' = {}", metric.name, value);
    
    match metric.name.as_str() {
        "claude.tokens.total" | "claude_code.token.usage" => {
            if let Some(token_type) = get_attribute(metric, "type")
                .or_else(|| get_attribute(metric, "token_type"))
            {
                println!("OTLP: Token type: {}, value: {}", token_type, value);
                match token_type.as_str() {
                    "input" => telemetry.tokens_input += value as u64,
                    "output" => telemetry.tokens_output += value as u64,
                    _ => println!("OTLP: Unknown token type: {}", token_type),
                }
                telemetry.tokens_total = telemetry.tokens_input + telemetry.tokens_output;
            }
        }
        "claude.api_cost.total" | "claude_code.cost.usage" => {
            println!("OTLP: Adding cost: ${}", value);
            telemetry.cost_usd += value;
        }
        "claude.active_time.duration" | "claude_code.active_time.ms" => {
            println!("OTLP: Active time: {}ms", value);
            telemetry.active_time_ms = value as u64;
        }
        "claude.tool_usage.total" | "claude_code.tool_usage.count" => {
            if let Some(tool_name) = get_attribute(metric, "tool")
                .or_else(|| get_attribute(metric, "tool_name"))
            {
                println!("OTLP: Tool '{}' used {} times", tool_name, value);
                *telemetry.tool_usage.entry(tool_name).or_insert(0) += value as u64;
            }
        }
        _ => println!("OTLP: Unhandled metric: {}", metric.name),
    }
}

fn get_metric_value(metric: &OtlpMetric) -> f64 {
    let binding = vec![];
    let data_points = metric
        .sum
        .as_ref()
        .or(metric.gauge.as_ref())
        .map(|m| &m.data_points)
        .unwrap_or(&binding);
    
    if let Some(point) = data_points.first() {
        point.as_double.unwrap_or_else(|| point.as_int.unwrap_or(0) as f64)
    } else {
        0.0
    }
}

fn get_attribute(metric: &OtlpMetric, key: &str) -> Option<String> {
    let binding = vec![];
    let data_points = metric
        .sum
        .as_ref()
        .or(metric.gauge.as_ref())
        .map(|m| &m.data_points)
        .unwrap_or(&binding);
    
    if let Some(point) = data_points.first() {
        if let Some(attrs) = &point.attributes {
            for attr in attrs {
                if attr.key == key {
                    return attr.value.string_value.clone()
                        .or_else(|| attr.value.int_value.map(|v| v.to_string()));
                }
            }
        }
    }
    
    None
}

fn process_log_attributes(telemetry: &mut TelemetryData, attrs: Vec<Attribute>) {
    let mut event_name = None;
    let mut input_tokens = 0u64;
    let mut output_tokens = 0u64;
    let mut cache_read_tokens = 0u64;
    let mut cache_creation_tokens = 0u64;
    let mut cost = 0.0;
    
    //println!("OTLP: Processing log attributes:");
    for attr in attrs {
        // println!("OTLP:   {} = {:?}{:?}", attr.key, attr.value.string_value, attr.value.int_value);
        match attr.key.as_str() {
            "event.name" => event_name = attr.value.string_value,
            "input_tokens" => {
                // Parse from string value since it's coming as a string
                if let Some(s) = &attr.value.string_value {
                    input_tokens = s.parse::<u64>().unwrap_or(0);
                } else {
                    input_tokens = attr.value.int_value.unwrap_or(0) as u64;
                }
            },
            "output_tokens" => {
                if let Some(s) = &attr.value.string_value {
                    output_tokens = s.parse::<u64>().unwrap_or(0);
                } else {
                    output_tokens = attr.value.int_value.unwrap_or(0) as u64;
                }
            },
            "cache_read_tokens" => {
                if let Some(s) = &attr.value.string_value {
                    cache_read_tokens = s.parse::<u64>().unwrap_or(0);
                } else {
                    cache_read_tokens = attr.value.int_value.unwrap_or(0) as u64;
                }
            },
            "cache_creation_tokens" => {
                if let Some(s) = &attr.value.string_value {
                    cache_creation_tokens = s.parse::<u64>().unwrap_or(0);
                } else {
                    cache_creation_tokens = attr.value.int_value.unwrap_or(0) as u64;
                }
            },
            "cost_usd" => {
                if let Some(s) = attr.value.string_value {
                    cost = s.parse::<f64>().unwrap_or(0.0);
                }
            }
            _ => {}
        }
    }
    
    if let Some(name) = event_name {
        println!("OTLP: Event '{}' - input: {}, output: {}, cache_read: {}, cache_creation: {}, cost: ${}", 
            name, input_tokens, output_tokens, cache_read_tokens, cache_creation_tokens, cost);
        if name == "api_request" || name == "user_prompt" {
            telemetry.tokens_input += input_tokens;
            telemetry.tokens_output += output_tokens;
            telemetry.tokens_total = telemetry.tokens_input + telemetry.tokens_output;
            telemetry.cost_usd += cost;
            
            // Track cache tokens separately
            if !telemetry.cache_read_tokens.is_some() {
                telemetry.cache_read_tokens = Some(0);
            }
            if !telemetry.cache_creation_tokens.is_some() {
                telemetry.cache_creation_tokens = Some(0);
            }
            *telemetry.cache_read_tokens.as_mut().unwrap() += cache_read_tokens;
            *telemetry.cache_creation_tokens.as_mut().unwrap() += cache_creation_tokens;
            
            println!("OTLP: Updated totals - tokens: {} (cache read: {}, cache creation: {}), cost: ${}", 
                telemetry.tokens_total, 
                telemetry.cache_read_tokens.unwrap_or(0),
                telemetry.cache_creation_tokens.unwrap_or(0),
                telemetry.cost_usd);
        }
    }
}

fn calculate_progress(telemetry: &TelemetryData) -> u8 {
    // Simple progress calculation based on tokens
    // Assume average PRP generation uses ~2000-5000 tokens
    let token_progress = (telemetry.tokens_total as f64 / 3000.0 * 100.0).min(90.0);
    token_progress as u8
}